File: Assets/src/model/actors/Player.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;
using UnityEngine;

[Serializable]
[ObjectInfo(description: "Only you can use and equip items.\nOnly you can take stairs.", flavorText: "Though your illness makes you physically weak, your knowledge of flora and fauna helps you navigate these strange caves.")]
public class Player : Actor, IBodyMoveHandler, IAttackHandler,
  ITakeAnyDamageHandler, IDealAttackDamageHandler, IActionPerformedHandler,
  IKillEntityHandler, IStatusAddedHandler, IHideInSidebar, IDeathHandler {
  // private float timeLastLostWater = 0;
  private int m_water;
  public int water {
    get => m_water;
    set {
      var diff = value - m_water;
      m_water = value;
      if (diff > 0) {
        GameModel.main.stats.waterCollected += diff;
      }
      OnChangeWater?.Invoke(diff);
    }
  }

  public bool isCamouflaged => Modifiers.Of<IPlayerCamouflage>(this).Any();

  // go to full HP and remove all debuffs
  internal void Replenish() {
    // doesn't count as a heal
    hp = maxHp;
    var debuffs = statuses.list.Where(s => s.isDebuff).ToArray();
    foreach (var debuff in debuffs) {
      statuses.Remove(debuff);
    }
  }

  internal readonly ItemHands Hands;
  [NonSerialized] /// lazily instantiated
  private HashSet<Actor> lastVisibleEnemies;
  public Inventory inventory { get; }
  public Equipment equipment { get; }

  public override IEnumerable<object> MyModifiers => base.MyModifiers.Concat(equipment);

  public override float turnPriority => 10;
  [field:NonSerialized] /// controller only, int delta
  public event Action<int> OnChangeWater;

  public Player(Vector2Int pos) : base(pos) {
    faction = Faction.Ally;
    inventory = new Inventory(10);
    // TODO the UX is broken
    // inventory.allowDragAndDrop = true;

    equipment = new Equipment(this);
    Hands = new ItemHands(this);
    hp = baseMaxHp = 12;
  }

  [OnSerializing]
  private void HandleSerializing(StreamingContext context) {
    /// clear player tasks on save
    ClearTasks();
  }

  public void HandleDeath(Entity source) {
    GameModel.main.GameOver(false, source);
  }

  public void HandleMove(Vector2Int newPos, Vector2Int oldPos) {
    floor.RecomputeVisibility();
    UpdateVisibleEnemies();
  }

  private void UpdateVisibleEnemies() {
    GameModel.main.EnqueueEvent(() => {
      var visibleEnemies = new HashSet<Actor>(GetVisibleActors(Faction.Enemy));
      // if there's a newly visible enemy from last turn, cancel the current move task
      var isNewlyVisibleEnemy = visibleEnemies.Any((enemy) => lastVisibleEnemies != null && !lastVisibleEnemies.Contains(enemy));
      if (isNewlyVisibleEnemy && task is FollowPathTask) {
        ClearTasks();
      }
      lastVisibleEnemies = visibleEnemies;
    });
  }

  public void HandleTakeAnyDamage(int damage) {
    GameModel.main.stats.damageTaken += damage;
    if (task is FollowPathTask) {
      ClearTasks();
    }
  }

  public void HandleDealAttackDamage(int damage, Body target) {
    GameModel.main.stats.damageDealt += damage;
  }

  public void OnKill(Entity entity) {
    if (entity is Actor a && a.faction == Faction.Enemy) {
      GameModel.main.stats.enemiesDefeated += 1;
    }
  }

  protected override void OnMoveFailed(Vector2Int target) {
    ClearTasks();
  }

  public void HandleActionPerformed(BaseAction final, BaseAction initial) {
    // GameModel.main.EnqueueEvent(() => {
    //   if (GameModel.main.time - timeLastLostWater > 10) {
    //     if (water > 0) {
    //       water -= 1;
    //     }
    //     timeLastLostWater = GameModel.main.time;
    //   }
    // });
    // player didn't do what they intended! We should reset and give
    // player a choice.
    if (final != initial) {
      ClearTasks();
    }
  }

  public void HandleStatusAdded(Status status) {
    if (status.isDebuff) {
      // cancel current action
      ClearTasks();
    }
  }

  protected override void HandleEnterFloor() {
    floor.RecomputeVisibility();
    UpdateVisibleEnemies();
    floor.timePlayerEntered = GameModel.main.time;
  }

  public void OnAttack(int damage, Body target) {
    var item = equipment[EquipmentSlot.Weapon];
    if (!(target is Destructible)) {
      if (item is IDurable durable) {
        GameModel.main.EnqueueEvent(durable.ReduceDurability);
      } else if (item is IStackable s) {
        GameModel.main.EnqueueEvent(() => s.stacks--);
      }
    }
    if (task is FollowPathTask) {
      task = null;
    }
  }

  internal override (int, int) BaseAttackDamage() {
    var item = equipment[EquipmentSlot.Weapon];
    if (item is IWeapon w) {
      return w.AttackSpread;
    } else {
      Debug.Log("Player attacking with a non-weapon in the weapon slot: " + item);
      return (1, 1);
    }
  }

  public IEnumerable<Tile> GetVisibleTiles() => floor.tiles.Where(t => t.isVisible);

  // Actors that are Visible
  public IEnumerable<Actor> GetVisibleActors(Faction faction) => floor.bodies.Where((b) => b is Actor a && a.isVisible && faction.HasFlag(a.faction)).Cast<Actor>();

  /// <summary>Return true if there are any enemies in vision range.</summary>
  public bool IsInCombat() {
    return GetVisibleActors(Faction.Enemy).Any();
  }
}

File: web/src/model/Player.ts
import { Actor, ATTACK_HANDLER, DEAL_ATTACK_DAMAGE_HANDLER, ACTION_PERFORMED_HANDLER, STATUS_ADDED_HANDLER, type IAttackHandler, type IDealAttackDamageHandler, type IActionPerformedHandler } from './Actor';
import { Body, BODY_MOVE_HANDLER, TAKE_ANY_DAMAGE_HANDLER, type IBodyMoveHandler, type ITakeAnyDamageHandler } from './Body';
import { Vector2Int } from '../core/Vector2Int';
import { Faction } from '../core/types';
import { collectModifiers } from '../core/Modifiers';
import { GameModelRef } from './GameModelRef';
import { Inventory } from './Inventory';
import { Equipment, EquipmentSlot } from './Equipment';
import { WEAPON_TAG, DURABLE_TAG, type IWeapon, type IDurable, reduceDurability } from './Item';
import { ItemHands } from './items/ItemHands';
import { EventEmitter } from '../core/EventEmitter';
import type { BaseAction } from './BaseAction';
import type { Entity } from './Entity';

const KILL_HANDLER = Symbol.for('IKillEntityHandler');
const DEATH_HANDLER = Symbol.for('IDeathHandler');
const CAMOUFLAGE = Symbol.for('IPlayerCamouflage');

/**
 * The player character.
 * Port of C# Player.cs.
 */
export class Player extends Actor {
  readonly inventory: Inventory;
  readonly equipment: Equipment;
  readonly hands: ItemHands;
  readonly onChangeWater = new EventEmitter<[number]>();

  get turnPriority(): number {
    return 10;
  }

  get isCamouflaged(): boolean {
    return collectModifiers<any>(this, CAMOUFLAGE).length > 0;
  }

  get myModifiers(): Iterable<object | null | undefined> {
    return [...super.myModifiers, ...this.equipment];
  }

  // Handler symbols — these mark Player as implementing these interfaces
  readonly [BODY_MOVE_HANDLER] = true;
  readonly [TAKE_ANY_DAMAGE_HANDLER] = true;
  readonly [ATTACK_HANDLER] = true;
  readonly [DEAL_ATTACK_DAMAGE_HANDLER] = true;
  readonly [ACTION_PERFORMED_HANDLER] = true;
  readonly [STATUS_ADDED_HANDLER] = true;
  readonly [DEATH_HANDLER] = true;
  readonly [KILL_HANDLER] = true;

  constructor(pos: Vector2Int) {
    super(pos);
    this.faction = Faction.Ally;
    this.inventory = new Inventory(10);
    this.equipment = new Equipment(this);
    this.hands = new ItemHands(this);
    this._hp = this._baseMaxHp = 12;
  }

  /** IDeathHandler */
  handleDeath(_source: Entity): void {
    GameModelRef.main.enqueuEvent(() => {
      // GameOver handled by GameModel
    });
  }

  /** IBodyMoveHandler */
  handleMove(_newPos: Vector2Int, _oldPos: Vector2Int): void {
    this.floor?.recomputeVisibility();
  }

  /** ITakeAnyDamageHandler */
  handleTakeAnyDamage(_damage: number): void {
    // Cancel path following on damage
    if (this.task && this.task.constructor.name === 'FollowPathTask') {
      this.clearTasks();
    }
  }

  /** IDealAttackDamageHandler */
  handleDealAttackDamage(_damage: number, _target: Body): void {
    // stats tracking
  }

  /** IKillEntityHandler */
  onKill(_entity: Entity): void {
    // stats tracking
  }

  /** IActionPerformedHandler */
  handleActionPerformed(finalAction: BaseAction, initialAction: BaseAction): void {
    if (finalAction !== initialAction) {
      this.clearTasks();
    }
  }

  /** IAttackHandler */
  onAttack(_damage: number, _target: Body): void {
    const item = this.equipment.get(EquipmentSlot.Weapon);
    if (item && DURABLE_TAG in item) {
      GameModelRef.main.enqueuEvent(() => reduceDurability(item as unknown as IDurable));
    }
    if (this.task && this.task.constructor.name === 'FollowPathTask') {
      this.task = null;
    }
  }

  baseAttackDamage(): [number, number] {
    const item = this.equipment.get(EquipmentSlot.Weapon);
    if (item && WEAPON_TAG in item) {
      return (item as unknown as IWeapon).attackSpread;
    }
    return [1, 1];
  }

  protected onMoveFailed(_target: Vector2Int): void {
    this.clearTasks();
  }

  protected handleEnterFloor(): void {
    this.floor?.recomputeVisibility();
    if (this.floor) {
      this.floor.timePlayerEntered = GameModelRef.main?.time ?? 0;
    }
  }

  replenish(): void {
    this._hp = this.maxHp;
    const debuffs = this.statuses.list.filter(s => s.isDebuff);
    for (const d of debuffs) {
      this.statuses.remove(d);
    }
  }

  getVisibleActors(faction: Faction): Actor[] {
    if (!this.floor) return [];
    const result: Actor[] = [];
    for (const b of this.floor.bodies) {
      if ('faction' in b && (b as Actor).isVisible && ((b as Actor).faction & faction) !== 0) {
        result.push(b as Actor);
      }
    }
    return result;
  }

  isInCombat(): boolean {
    return this.getVisibleActors(Faction.Enemy).length > 0;
  }
}
